src/diffKernel.jl

```diff
@@ -0,0 +1,81 @@
+using OneHotArrays: OneHotVector
+import ForwardDiff as FD
+import LinearAlgebra as LA
+
+""" 
+	DiffPt(x; partial=())
+
+For a covariance kernel k of GP Z, i.e.
+```julia
+	k(x,y) # = Cov(Z(x), Z(y)),
+```
+a DiffPt allows the differentiation of Z, i.e.	
+```julia
+	k(DiffPt(x, partial=1), y) # = Cov(∂₁Z(x), Z(y))
+```
+for higher order derivatives partial can be any iterable, i.e.
+```julia
+	k(DiffPt(x, partial=(1,2)), y) # = Cov(∂₁∂₂Z(x), Z(y))
+```
+"""
+struct DiffPt{Dim}
+    pos # the actual position
+    partial
+end
+
+DiffPt(x; partial=()) = DiffPt{length(x)}(x, partial) # convenience constructor
+
+"""
+Take the partial derivative of a function `fun`  with input dimesion `dim`.
+If partials=(i,j), then (∂ᵢ∂ⱼ fun) is returned.
+"""
+function partial(fun, dim, partials=())
+    if !isnothing(local next = iterate(partials))
+        idx, state = next
+        return partial(
+            x -> FD.derivative(0) do dx
+                fun(x .+ dx * OneHotVector(idx, dim))
```

Comment:
Why `OneHotVector`? Can't you just use
```suggestion
                y = similar(x)
                copyto!(y, x)
                y[idx] += dx
                fun(y)
```
Using `OneHotVector` does not seem to save any allocations here.

Reply:
I wanted to avoid mutation as autodiff seems to be unstable if you do use mutation. In principle the code above is probably equivalent - as long as autodiff does not stumble over it.

Reply:
This is also the only reason OneHotArrays was used. If that truly works with mutation this dependency could be removed.

Reply:
> autodiff seems to be unstable if you do use mutation.

Definitely not the case for ForwardDiff.

For immutable arrays you would probably just dispatch/branch above and then use `y = Base.setindex(x, x[idx] + dx, idx)` - a bit unfortunate that Base does not support this for `Array`. But then again usually we only work with `Array`, so this special case is not even needed (definitely not initially).

Reply:
the `similar(x)` thing breaks if `x` is just a value not a vector :(

Reply:
Simple fix is to dispatch on `x` being a vector or a scalar.